### PR TITLE
[FEATURE] support deb/rpm/exe/dmg packages and SSE4+POPCNT binaries

### LINKED ISSUE #1667: Better packages
CPack can generate `.dmg` for Mac; `.deb` and `.rpm` for Linux and proper installers for Windows. Why aren't we doing this? Also we should have optimized builds and unoptimized.

i.e.

**Yara**

Version  | Linux                                          | Mac                              | FreeBSD          | Windows
----------  | ----------------------------------------- | -------------------------------| -------------------- | -------------
0.9.6      | 32bit: [tar.xz]() [deb]() [rpm]()     | 32bit: [tar.xz]() [dmg]() | 32bit: [tar.xz]()  | 32bit: [zip]() [exe]()
              | 64bit: [tar.xz]() [deb]() [rpm]()     | 64bit: [tar.xz]() [dmg]() | 64bit: [tar.xz]()  | 64bit: [zip]() [exe]()
              | sse4: [tar.xz]() [deb]() [rpm]()     | sse4: [tar.xz]() [dmg]() | sse4: [tar.xz]()  | sse4: [zip]() [exe]()
0.9.5      | 32bit: [tar.xz]() [deb]() [rpm]()     | 32bit: [tar.xz]() [dmg]() | 32bit: [tar.xz]()  | 32bit: [zip]() [exe]()
              | 64bit: [tar.xz]() [deb]() [rpm]()     | 64bit: [tar.xz]() [dmg]() | 64bit: [tar.xz]()  | 64bit: [zip]() [exe]()
              | sse4: [tar.xz]() [deb]() [rpm]()     | sse4: [tar.xz]() [dmg]() | sse4: [tar.xz]()  | sse4: [zip]() [exe]()

### PR DESCRIPTION
**Pull request by xenigmax**
1.
Below packages will be made.
1) .exe for Windows (tested on Klette)
2) .dmg for Mac (tested on cyclops)
3) .deb/rpm for linux/unix (tested on giraffe)

2.
If a user set -DSEQAN_ARCH_SSE4=1, 
1) -msse4 and -mpopcnt will be applied.
2) Binary file name will be *x86_64_sse4* instead of x86_64 (tested on Klette, cyclops, giraffe)

This will close #1667

### CODE REVIEW

File: util/cmake/SeqAnBuildSystem.cmake [hunk 1]
```diff
@@ -437,12 +437,20 @@ macro (seqan_configure_cpack_app APP_NAME APP_DIR)
   # The following include automates the MS Redistributable installer.
   include (InstallRequiredSystemLibraries)
 
+  SET(CPACK_GENERATOR "ZIP")
```
Comment (h-2) [inline util/cmake/SeqAnBuildSystem.cmake]: We only need ZIP for mac and windows, not linux and bsd (they have the tarballs).

File: util/cmake/SeqAnBuildSystem.cmake [hunk 2]
```diff
@@ -437,12 +437,20 @@ macro (seqan_configure_cpack_app APP_NAME APP_DIR)
...
   else ()
-    set (CPACK_GENERATOR "ZIP;TXZ")
+    if (CMAKE_VERSION VERSION_LESS "3.1") # TXZ support since 3.1
+      SET(CPACK_GENERATOR "${CPACK_GENERATOR};TBZ2")
+    else()
+      SET(CPACK_GENERATOR "${CPACK_GENERATOR};TXZ")
+    endif ()
+    if (CMAKE_SYSTEM_NAME MATCHES "Darwin")
+      SET(CPACK_GENERATOR "${CPACK_GENERATOR};DragNDrop")
+    else ()
```
Comment (h-2) [inline util/cmake/SeqAnBuildSystem.cmake]: add Linux as condition, since on BSD we don't want to build deb and rpm.

**Comment by h-2** [comment]
For the generators, I think the following pseudocode would be best:
```
if (windows)
  ZIP;NSIS
elseif (darwin)
  ZIP;DragNDrop
elseif (cmake_version < 3.1)
 TBZ2
else
 TXZ
endif

if (linux)
 += DEB,RPM
endif
```

For the architecture name we now have SEQAN_ARCH_AVX2, as well, can you add that, too? And check it **before** checking SEQAN_ARCH_SSE4?

Thanks!

File: util/cmake/SeqAnBuildSystem.cmake (hunk 1 above)
Comment (xenigmax) [inline util/cmake/SeqAnBuildSystem.cmake]: According to the package site, we provide zip files for linux and bsd as well.
http://packages.seqan.de/razers3/
http://packages.seqan.de/yara/

Would you like to change this?
I will change it if you confirm this.

Thanks~!

**Comment by xenigmax** [comment]
@h-2 
According to the package site, we provide zip files for linux and bsd as well.
http://packages.seqan.de/razers3/
http://packages.seqan.de/yara/

Would you like to change this?
I will change it if you confirm this.

Thanks~!

**Comment by h-2** [comment]
> Would you like to change this?

Yes, since every linux distribution supports tarballs out of the box (but not zip), I see no reason that we should provide zip for linux.

File: util/cmake/SetCPackSystemName.cmake [hunk 3]
```diff
@@ -44,8 +44,12 @@ if (CMAKE_SYSTEM_PROCESSOR STREQUAL "amd64")
   set(CMAKE_SYSTEM_PROCESSOR "x86_64")
 endif ()
 
-if ((CMAKE_SYSTEM_PROCESSOR STREQUAL "x86_64") AND (SEQAN_ARCH_SSE4))
-  set (CMAKE_SYSTEM_PROCESSOR "x86_64_sse4") 
+if (CMAKE_SYSTEM_PROCESSOR STREQUAL "x86_64" AND (SEQAN_ARCH_SSE4))
```
Comment (xenigmax) [inline util/cmake/SetCPackSystemName.cmake]: @h-2 
I made this according to https://github.com/seqan/seqan/blob/develop/util/cmake/SeqAnBuildSystem.cmake#L287
However, correct me if it's not what you wanted!

File: util/cmake/SetCPackSystemName.cmake [hunk 4]
```diff
@@ -44,8 +44,12 @@ if (CMAKE_SYSTEM_PROCESSOR STREQUAL "amd64")
...
+  if (SEQAN_ARCH_AVX2)
+    set (CMAKE_SYSTEM_PROCESSOR "x86_64_avx2") 
+  else ()
+    set (CMAKE_SYSTEM_PROCESSOR "x86_64_sse4") 
+  endif ()
```
Comment (xenigmax) [inline util/cmake/SetCPackSystemName.cmake]: @h-2 
I made this according to https://github.com/seqan/seqan/blob/develop/util/cmake/SeqAnBuildSystem.cmake#L287
However, correct me if it's not what you wanted. :)

Comment (h-2) [inline util/cmake/SetCPackSystemName.cmake]: perfect :+1:

**Comment by h-2** [comment]
I will wait for the CI-build and then merge this. If I am not mistaken, the package-site build-script still needs changes, but they will be done in http://github.com/seqan/packages.seqan.de

**Comment by xenigmax** [comment]
Yes. I will do that in that repository. Don't worry :)

**Pull request merged**